### PR TITLE
8224 report required permissions when authorization fails

### PR DESCRIPTION
**Pull request by ItamarYuran**
Closes #(8224)

## Change Description
The added code helps the user understand why they are not authorized to perform a certain action.

### Background
Up until now in case a user had no permissions or was denied of some action, they would only get a `401 unauthorized` message, making it hard to understand what is missing in order to perform that action.
      
The enhancement will report all denied actions in case there are any, or if there aren't any - report missing permissions.

### Testing Details
Changes were not tested yet, no existing test got broken

## Additional info

Screenshot 2024-10-28 at 13 59 02
Screenshot 2024-10-28 at 14 02 29

### CODE REVIEW

**Review by arielshaqed — COMMENTED** [review]
Nice!  Can you please show what outputs we can expect?  Or is that in the next commit with tests etc.?

**Comment by ItamarYuran** [comment]
The current code returns:
 - All denied actions in a policy in case there are any
 - First missing permission in case there is no relevant denied action

File: pkg/auth/service.go [hunk 1]
```diff
@@ -1147,10 +1152,32 @@ func NewAPIAuthServiceWithClient(client ClientWithResponsesInterface, externalPr
...
+		return strings.TrimSpace(w.String())
+	}
+	if len(n.Unauthorized) != 0 {
+		fmt.Fprintln(w, "lacking permissions for:")
+		for _, node := range n.Unauthorized {
+			fmt.Fprintln(w, node.Permission.Action)
+		}
+		return strings.TrimSpace(w.String())
+	}
+	return strings.TrimSpace(w.String())
```
Comment (arielshaqed) [inline pkg/auth/service.go:1173]: This returns `""`.  If that is not helpful, please return some constant string like "not allowed".

File: pkg/auth/service.go [hunk 2]
```diff
@@ -44,6 +44,11 @@ type AuthorizationResponse struct {
 	Error   error
 }
 
+type NeededPermissions struct {
+	Denied       []model.Statement
+	Unauthorized []permissions.Node
```
Comment (arielshaqed) [inline pkg/auth/service.go:49]: These are a bit too informative IMO.  Let's just keep the action name here?  Otherwise I learn details of the policy _including denied paths that I might not have known existed!_
```suggestion
	// Denied is a list of actions the user was denied for the attempt.
	Denied       []string
	// Unauthorized is a list of actions the user did not have for the attempt.
	Unauthorized []string
```

File: pkg/auth/service.go [hunk 3]
```diff
@@ -1147,10 +1152,32 @@ func NewAPIAuthServiceWithClient(client ClientWithResponsesInterface, externalPr
...
+		fmt.Fprintln(w, "lacking permissions for:")
+		for _, node := range n.Unauthorized {
+			fmt.Fprintln(w, node.Permission.Action)
+		}
+		return strings.TrimSpace(w.String())
+	}
+	return strings.TrimSpace(w.String())
+}
+
+func CheckPermissions(ctx context.Context, node permissions.Node, username string, policies []*model.Policy, perm *NeededPermissions) CheckResult {
```
Comment (Isan-Rivkin) [inline pkg/auth/service.go:1176]: nit: variable name `perm` is too generic. 
Can you use something more explicit? 
for example: `audit`, `denyAudit`, `permAudit`, `history` etc...
So that in the future when extending, debugging or just reading the code we can immediately understand what it is used for.

File: pkg/auth/service.go [hunk 4]
```diff
@@ -1165,23 +1189,28 @@ func CheckPermissions(ctx context.Context, node permissions.Node, username strin
 
 					if stmt.Effect == model.StatementEffectDeny {
 						// this is a "Deny" and it takes precedence
-						return CheckDeny
+						perm.Denied = append(perm.Denied, stmt)
```
[the remark this answers is not in the document]
Comment (Isan-Rivkin) [reply pkg/auth/service.go:1195]: Was thinking the same, agree with @arielshaqed 100%

File: pkg/auth/service.go [hunk 5]
```diff
@@ -44,6 +44,11 @@ type AuthorizationResponse struct {
 	Error   error
 }
 
+type NeededPermissions struct {
```
Comment (Isan-Rivkin) [inline pkg/auth/service.go:47]: rename, `MissingPermissions` or `PermissionsAudit` or `DenyHistory` w/e would be more appropriate?

File: contrib/auth/acl/service.go [hunk 6]
```diff
@@ -895,8 +895,8 @@ func (s *AuthService) Authorize(ctx context.Context, req *auth.AuthorizationRequ
 	if err != nil {
 		return nil, err
 	}
-
-	allowed := auth.CheckPermissions(ctx, req.RequiredPermissions, req.Username, policies)
+	perm := &auth.NeededPermissions{}
```
Comment (Isan-Rivkin) [inline contrib/auth/acl/service.go:898]: sorry, Im confused what are you doing with this variable?

**Review by Isan-Rivkin — CHANGES_REQUESTED** [review]

**Comment by ItamarYuran** [comment]
Thank you for your reviews!
These changes will return all denied permissions, only the ones that regard the request (not all permission in the policy)
Same goes for missing permissions, all missing one will return. In case of denied + missing permissions only denied ones return.

File: pkg/auth/service.go [hunk 7]
```diff
@@ -1147,10 +1155,21 @@ func NewAPIAuthServiceWithClient(client ClientWithResponsesInterface, externalPr
 	}, nil
 }
 
-func CheckPermissions(ctx context.Context, node permissions.Node, username string, policies []*model.Policy) CheckResult {
+func (n *MissingPermissions) String() string {
+	if len(n.Denied) != 0 {
+		return fmt.Sprintf("denied permission to %s", strings.Join(n.Denied, ","))
+	}
+	if len(n.Unauthorized) != 0 {
+		return fmt.Sprintf("missing permission to %s", strings.Join(n.Unauthorized, ","))
```
Comment (arielshaqed) [inline pkg/auth/service.go:1163]: l.935 will cause a message that says `"insufficient permissions: missing permission to ..."`.  That's a bit wordy, so perhaps
```suggestion
		return fmt.Sprintf("not allowed to %s", strings.Join(n.Unauthorized, ","))
```
An added benefit is that it uses the same word "allow" from the policy language.

File: pkg/auth/service.go [hunk 8]
```diff
@@ -1165,21 +1184,26 @@ func CheckPermissions(ctx context.Context, node permissions.Node, username strin
 
 					if stmt.Effect == model.StatementEffectDeny {
 						// this is a "Deny" and it takes precedence
+						permAudit.Denied = append(permAudit.Denied, action)
 						return CheckDeny
+					} else {
+						hasPermission = true
+						allowed = CheckAllow
```
Comment (arielshaqed) [inline pkg/auth/service.go:1191]:
```suggestion
					hasPermission = true
					allowed = CheckAllow
```
is closer to the previous version.

We prefer to use an "early return" style: errors return immediately - like l. 1188 does; successes keep going with no "else" and no need to indent.

**Review by arielshaqed — APPROVED** [review]
Very elegant, thanks!  I think this will be a great improvement in quality of life for lakeFS admins who need to configure permissions.

Just a note about phrasing one of the error messages; please decide if it will improve things and pull either way.

File: contrib/auth/acl/service.go [hunk 9]
```diff
@@ -895,8 +895,8 @@ func (s *AuthService) Authorize(ctx context.Context, req *auth.AuthorizationRequ
 	if err != nil {
 		return nil, err
 	}
-
-	allowed := auth.CheckPermissions(ctx, req.RequiredPermissions, req.Username, policies)
+	perm := &auth.MissingPermissions{}
```
Comment (Isan-Rivkin) [inline contrib/auth/acl/service.go:898]: Going back to prev question - what are you doing with `perm` after `auth.CheckPermissions` ? I don't see it being used.

**Review by Isan-Rivkin — APPROVED** [review]
single comment regarding ACL authorize check. its not used there. 
Don't want to block PTAL
